framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,626 @@
+        String methodName = aMethod.methodName.substring(0, aMethod.methodName.indexOf("("));
+        // Use Java syntax for constructors.
+        if ("<init>".equals(methodName)) {
+            // Constructor names cannot contain dots, if this is an inner class.
+            methodName =
+                    basename.contains(".")
+                            ? basename.substring(basename.lastIndexOf('.') + 1)
+                            : basename;
+        } else {
+            // This isn't a constructor, so add a return type.
+            // Note that the stub file format doesn't require this to be correct,
+            // so it would be acceptable to print "java.lang.Object" for every
+            // method. A better type is printed if one is available to improve
+            // the readability of the resulting stub file.
+            printWriter.print(formatAnnotations(aMethod.returnType.tlAnnotationsHere));
+            String returnType = aMethodWrapper.getReturnType();
+            printWriter.print(returnType);
+            printWriter.print(" ");
+        }
+        printWriter.print(methodName);
+        printWriter.print("(");
+
+        StringJoiner parameters = new StringJoiner(", ");
+
+        if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()) {
+            // Only output the receiver if it has an annotation.
+            parameters.add(
+                    formatParameter(
+                            AFieldWrapper.createReceiverParameter(aMethod.receiver, basename),
+                            "this",
+                            basename));
+        }
+        for (Integer index : aMethodWrapper.getParameters().keySet()) {
+            AFieldWrapper param = aMethodWrapper.getParameters().get(index);
+            parameters.add(formatParameter(param, param.getParameterName(), basename));
+        }
+        printWriter.print(parameters.toString());
+        printWriter.println(");");
+        printWriter.println();
+    }
+
+    /**
+     * The implementation of {@link #write(ASceneWrapper, Writer)} and {@link #write(ASceneWrapper,
+     * String)}. Prints imports, classes, method signatures, and fields in stub file format, all
+     * with appropriate annotations.
+     *
+     * @param scene the scene to write
+     * @param printWriter where to write the scene to
+     */
+    private static void writeImpl(ASceneWrapper scene, PrintWriter printWriter) {
+        // Write out all imports
+        ImportDefWriter importDefWriter;
+        try {
+            importDefWriter = new ImportDefWriter(scene, printWriter);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+        importDefWriter.visit();
+        printWriter.println();
+
+        // For each class
+        for (Map.Entry<@BinaryName String, AClassWrapper> classEntry :
+                scene.getClasses().entrySet()) {
+            printClass(classEntry.getKey(), classEntry.getValue(), printWriter);
```

Comment:
`printClass` behaves incorrectly if a class in the default package is processed after a class in any other package.  Therefore, the client code here should sort the classes by package, to put those in the default package first.  For determinism (which will make comparisons easier, such as during debugging), I suggest to sort by package name (default package first) then by class name within package.